### PR TITLE
#170947673  allow requester to be remembered nextime

### PR DESCRIPTION
**Pull request by EugeneMunya**
#### What does this PR do?
- Add remember me functionality
#### Description of Task to be completed?
- A requester will tick remember me checkbox and her/his profile information will be stored in browser sessions so that next time they can be used to create request.
#### How should this be manually tested?
- open your terminal and run this command `git clone https://github.com/andela/knightsbn    backend.git` to clone repository
- run `cd knights-bn-backend` to get inside the project
- run `git fetch && git checkout ft-remember-me-170947673` 
- run `npm install` to install dependencies
- check if you have PostgresQL database if you don't have it install it from here [PostgreSQL](https://www.2ndquadrant.com/en/blog/pginstaller-install-postgresql/)
- create database
- create `.env` file in your project root directory
- follow `env.example` file to create environment variables in `.env` file and 
   assign them with all your database configurations
- run `npm run createTables` to create tables in your new database you have created.
- run `npm run test` to run all the tests
- run `npm run dev-start` to start the server.
- open your postman or any other API tester you have
-  to test this end point you have to be logged in or signed up , see how to do [log in](https://github.com/andela/knights-bn-backend/pull/23) or [sign up](https://github.com/andela/knights-bn-backend/pull/17)
- send a GET request to this url `http://localhost:4000/api/v1/remembered`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#170947673](https://www.pivotaltracker.com/story/show/170947673)
#### Screenshots

Screen Shot 2020-03-03 at 21 58 18

### CODE REVIEW

**Comment by T2Wil** [comment]
Hey @EugeneMunya , I followed the PR description but I got different response after hitting Get http://localhost:4000/api/v1/remembered. Please check where the problem is.

image

**Comment by Niyonsengaeric** [comment]
@EugeneMunya me also I get different results 
Screen Shot 2020-03-03 at 16 50 59

File: src/tests/rememberMe.test.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+chai.should();
+
+const rememberMe = () => {
+  describe('Remember user who ckecked to be remembered ', () => {
+    it('it should return 200 when profile information strored in cookies successfull  ', (done) => {
+      chai
+        .request(app)
+        .get('/api/v1/remembered')
+        .end((err, res) => {
+          expect(res.statusCode).to.equal(200);
```
Comment (T2Wil) [inline src/tests/rememberMe.test.js:28]: You can also test data in cookies to make sure the `req.cookies` is not empty or anything.

File: src/tests/rememberMe.test.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+          expect(res.body.data).to.be.an('object');
+        });
+      done();
+    });
+
+    it('it should return 500 you are not in the database ', (done) => {
```
Comment (T2Wil) [inline src/tests/rememberMe.test.js]: I think `404` is a correct status for a not found user case.

Comment (EugeneMunya) [reply src/tests/rememberMe.test.js]: That mistake has been resolved, you check

File: src/tests/rememberMe.test.js (hunk 1 above)
Comment (EugeneMunya) [reply src/tests/rememberMe.test.js:28]: I have changed from storing in cookies to sessions and I managed to test it,  you can check for now

File: src/tests/rememberMe.test.js (hunk 2 above)
Comment (EugeneMunya) [reply src/tests/rememberMe.test.js]: > @EugeneMunya me also I get different results
> Screen Shot 2020-03-03 at 16 50 59
The issue was the time it takes to update cookie storage and I manage this by storing the user information in sessions check how I did it

**Comment by EugeneMunya** [comment]
> Hey @EugeneMunya , I followed the PR description but I got different response after hitting Get http://localhost:4000/api/v1/remembered. Please check where the problem is.
> 
> image

I forgot to update user the image in description now I store user information in the session and I return the user data stored in session

File: src/tests/rememberMe.test.js (hunk 1 above)
Comment (T2Wil) [reply src/tests/rememberMe.test.js:28]: Well done @EugeneMunya

**Review by T2Wil — APPROVED** [review]

File: src/app.js [hunk 3]
```diff
@@ -20,6 +25,16 @@ app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 multer({ dest: 'uploads/' });
 app.use('/uploads', express.static('uploads'));
+app.use(cookieParser());
+app.use(sessions({
+  secret: process.env.SECRETKEY,
+  resave: true,
+  saveUninitialized: false,
+  cookie: {
+    maxAge: 1000 * 365 * 24 * 60 * 60 * 2,
```
Comment (kagabof) [inline src/app.js]: Please reduce the session expiration time.

File: src/controllers/userProfile.js [hunk 4]
```diff
@@ -23,4 +23,15 @@ export default class changeUserProfile {
         response.status(200).json({ status: 200, user: returnProfile(user) });
       }).catch((dbError) => response.status(500).json({ error: 'Database issues', dbError }));
   }
+
+  static async rememberMe(req, res) {
+    try{
+   const userInfo = await models.User.findOne({ where: { id: req.user.id } })
+   const {password, ...rest} = userInfo.dataValues;
+   req.session.profileInfo = rest
+    return res.status(200).send({status:200, message:'Your profile information was saved for next use', data:req.session.profileInfo})
```
Comment (kagabof) [inline src/controllers/userProfile.js]: please change the responce message to make it more understandable.

**Review by kagabof — CHANGES_REQUESTED** [review]

File: src/controllers/userProfile.js (hunk 4 above)
Comment (EugeneMunya) [reply src/controllers/userProfile.js]: I have reduced the expiration date

**Review by kagabof — APPROVED** [review]
Great job.

**Pull request merged**